group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -0,0 +1,575 @@
+import org.apache.kafka.image.MetadataImage;
+import org.apache.kafka.server.record.BrokerCompressionType;
+import org.apache.kafka.server.util.FutureUtils;
+import org.slf4j.Logger;
+
+import java.util.List;
+import java.util.OptionalInt;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.IntSupplier;
+
+/**
+ * The group coordinator service.
+ */
+public class GroupCoordinatorService implements GroupCoordinator {
+
+    public static class Builder {
+        private final int nodeId;
+        private final GroupCoordinatorConfig config;
+        private PartitionWriter<Record> writer;
+        private CoordinatorLoader<Record> loader;
+
+        public Builder(
+            int nodeId,
+            GroupCoordinatorConfig config
+        ) {
+            this.nodeId = nodeId;
+            this.config = config;
+        }
+
+        public Builder withWriter(PartitionWriter<Record> writer) {
+            this.writer = writer;
+            return this;
+        }
+
+        public Builder withLoader(CoordinatorLoader<Record> loader) {
+            this.loader = loader;
+            return this;
+        }
+
+        public GroupCoordinatorService build() {
+            if (config == null)
+                throw new IllegalArgumentException("Config must be set.");
+            if (writer == null)
+                throw new IllegalArgumentException("Writer must be set.");
+            if (loader == null)
+                throw new IllegalArgumentException("Loader must be set.");
+
+            String logPrefix = String.format("GroupCoordinator id=%d ", nodeId);
+            LogContext logContext = new LogContext(String.format("[%s ]", logPrefix));
+
+            CoordinatorBuilderSupplier<ReplicatedGroupCoordinator, Record> supplier = () ->
+                new ReplicatedGroupCoordinator.Builder(config);
+
+            CoordinatorEventProcessor processor = new MultiThreadedEventProcessor(
+                logContext,
+                "group-coordinator-event-processor-",
+                config.numThreads
+            );
+
+            CoordinatorRuntime<ReplicatedGroupCoordinator, Record> runtime =
+                new CoordinatorRuntime.Builder<ReplicatedGroupCoordinator, Record>()
+                    .withLogPrefix(logPrefix)
+                    .withLogContext(logContext)
+                    .withEventProcessor(processor)
+                    .withPartitionWriter(writer)
+                    .withLoader(loader)
+                    .withCoordinatorBuilderSupplier(supplier)
+                    .build();
+
+            return new GroupCoordinatorService(
+                logContext,
+                config,
+                runtime
+            );
+        }
+    }
+
+    /**
+     * The name of the consumer metadata topic.
+     */
+    private static final String CONSUMER_METADATA_TOPIC_NAME = "__consumer_offsets";
```

Comment:
nit

I assume that this is going to be the source of truth for the topic name? In which case, we should probably use this constant to update `Topic.GROUP_METADATA_TOPIC_NAME`.

Reply:
Good point but I think that it should be the other way around. `Topic.GROUP_METADATA_TOPIC_NAME` should be the source of truth.